packages/blockchain/src/index.ts

```diff
@@ -460,12 +467,21 @@ export default class Blockchain implements BlockchainInterface {
         }
       }
     }
+
+    // trim length to CLIQUE_SIGNER_HISTORY_BLOCK_LIMIT
+    const length = this._cliqueLatestVotes.length
+    const limit = this.CLIQUE_SIGNER_HISTORY_BLOCK_LIMIT
+    if (length > limit) {
+      this._cliqueLatestVotes = this._cliqueLatestVotes.slice(length - limit, length)
+    }
```

Comment:
right now all votes are being cleared on epoch transitions: https://github.com/ethereumjs/ethereumjs-monorepo/blob/6721321fb46e8543eeb1027d11fdd8cdf8841994/packages/blockchain/src/index.ts#L822-L826

so I can implement the above logic you suggested instead. 

Is maybe 50 blocks a safe enough waiting period? I have no personal experience to get a feel for how long reorgs could possibly get on poa chains.

Reply:
I don't actually think it's too tricky: if there is a reorg, then all vote and signer state entries with block numbers higher than the reorg block number will be deleted, the signers from the  last singer state stack item will be become the new active signers. Everything else from that moment on the continues (restarts) as before.

If I don't get this wrong, votes need to be kept relatively long in this scenario, so i think "EPOCH_LENGTH + REORG_SAFE_PERIOD" block numbers - so something like 30050, since if a reorg is happening from let's say 30020 -> 29950 and there is voting taking place in these last 50 blocks before the epoch transition, theoretically all existing vote from this 0-29999 epoch might be relevant for the pre-epoch voting. Does this make sense?

For Signer States I would think - all states from the last REORG_SAFE_PERIOD blocks + minimally one state - should be enough if I am not forgetting something.

But I think we can also generally be a bit more generous here - e.g. take a somewhat longer REORG_SAFE_PERIOD of 100 or something or keep more signer states, nothing lost here, this doesn't take that much space and eventually the date might be interesting for analytical purposes.